dynet/cudnn-ops.cc

```diff
@@ -0,0 +1,227 @@
+    stride[i] = static_cast<int>(s[i]);
+  }
+  is_valid = padding_type;
+
+  fwd_workspace = NULL;
+  bwd_filter_workspace = NULL;
+  bwd_data_workspace = NULL;
+  workspace_fwd_size_ = 0;
+  workspace_bwd_data_size_ = 0;
+  workspace_bwd_filter_size_ = 0;
+  mempool_ = NULL;
+
+  CUDNN_CHECK(cudnnCreateTensorDescriptor(&x_desc_));
+  CUDNN_CHECK(cudnnCreateTensorDescriptor(&y_desc_));
+  CUDNN_CHECK(cudnnCreateTensorDescriptor(&bias_desc_));
+  CUDNN_CHECK(cudnnCreateFilterDescriptor(&filter_desc_));
+  CUDNN_CHECK(cudnnCreateConvolutionDescriptor(&conv_desc_));
+}
+
+CudnnConvOp::~CudnnConvOp() {
+  CUDNN_CHECK(cudnnDestroyTensorDescriptor(x_desc_));
+  CUDNN_CHECK(cudnnDestroyTensorDescriptor(y_desc_));
+  CUDNN_CHECK(cudnnDestroyTensorDescriptor(bias_desc_));
+  CUDNN_CHECK(cudnnDestroyFilterDescriptor(filter_desc_));
+  CUDNN_CHECK(cudnnDestroyConvolutionDescriptor(conv_desc_));
+}
+
+void CudnnConvOp::forward_impl(const Device_GPU & dev, const std::vector<const Tensor*>& xs, Tensor& fx) {
+  const Tensor* x = xs[0]; 
+  const Tensor* filter = xs[1];
+  Tensor* y = &fx;
+ 
+  unsigned XN = x->d.bd;
+  unsigned XC = x->d[2];
+  unsigned XH = x->d[0];
+  unsigned XW = x->d[1];
+  unsigned FYC = filter->d[3];
+  unsigned FXC = filter->d[2];
+  unsigned FH = filter->d[0];
+  unsigned FW = filter->d[1];
+  unsigned YN = fx.d.bd;
+  unsigned YC = fx.d[2];
+  unsigned YH = fx.d[0];
+  unsigned YW = fx.d[1];
+
+  // infer pad_h, pad_w
+  if (!is_valid) {
+  // Total padding on rows and cols is
+  // Pr = (R' - 1) * S + Kr - R
+  // Pc = (C' - 1) * S + Kc - C
+  // where (R', C') are output dimensions, (R, C) are input dimensions, S
+  // is stride, (Kr, Kc) are filter dimensions.
+  // We pad Pr/2 on the left and Pr - Pr/2 on the right, Pc/2 on the top
+  // and Pc - Pc/2 on the bottom.  When Pr or Pc is odd, this means
+  // we pad more on the right and bottom than on the top and left.
+    pad_h = std::max<int>(0, (YH - 1) * stride[0] + FH - XH);
+    pad_w = std::max<int>(0, (YW - 1) * stride[1] + FW - XW);
+    if (mempool_ == NULL) {
+      throw std::runtime_error("dynet::CudnnConvOp::mempool_ not set");
+    }
+    const bool h_odd = (pad_h % 2 != 0);
+    const bool w_odd = (pad_w % 2 != 0);
+    if (h_odd || w_odd) { // then we need to pad one row/col on the bottom/right
+      unsigned new_XH = XH + h_odd;
+      unsigned new_XW = XW + w_odd;
+      void* temp = mempool_->allocate(sizeof(float) * new_XW * new_XH * XC * XN);
+      padded_x = Tensor(Dim({new_XH, new_XW, XC}, XN), static_cast<float*>(temp), xs[0]->device, DeviceMempool::FXS);
+      //Eigen::array<std::pair<int, int>, 4> paddings;
+      //paddings[0] = std::make_pair(0, static_cast<int>(h_odd));
+      //paddings[1] = std::make_pair(0, static_cast<int>(w_odd));
+      //paddings[2] = std::make_pair(0, 0);
+      //paddings[3] = std::make_pair(0, 0);
+      //TODO this function cannot be linked
+      //padded_x.tb<3>().device(*dev.edevice) = xs[0]->tb<3>().pad(paddings);
```

Comment:
Here I tried to use Eigen custom kernels to do positive padding on GPU tensors but ended up with a linking error. Can someone familiar with Eigen GPU operations take a look?